### PR TITLE
Bluetooth details

### PR DESCRIPTION
**Pull request by ski-sketch**
### Prerequisites

- [x] I have [searched](https://github.com/DefinetlyNotAI/Logicytics/pulls) for duplicate or closed issues.
- [x] I have read the [contributing guidelines](https://github.com/DefinetlyNotAI/Logicytics/blob/main/CONTRIBUTING.md).
- [x] I have followed the instructions in the [wiki](https://github.com/DefinetlyNotAI/Logicytics/wiki) about contributions.
- [x] I have updated the documentation accordingly, if required.
- [x] I have tested my code with the `--dev` flag, if required.

### PR Type

- [x] Bug fix 
- [x] New feature 
- [x] Refactoring 
- [ ] Documentation
  update 
- [ ] ⚠️ Breaking change ⚠️ 

### Description

    Retrieves and logs detailed information about Bluetooth devices on the system.

    The function connects to the Windows Management Instrumentation (WMI) service and queries for devices
    whose names contain the term 'Bluetooth'. It writes the information to a text file named 'Bluetooth Info.txt'.

    Information for each device includes:
    - Name
    - Device ID
    - Description
    - Manufacturer
    - Status
    - PNP Device ID

### Motivation and Context

I thought that giving logicytics bluetooth capabilities would be a really good addition.

### Credit

_N/A_

### Issues Fixed

_N/A_

## Summary by CodeRabbit

- **New Features**
	- Introduced a script to retrieve and log detailed information about Bluetooth devices on Windows.
	- Added functionality for collecting and logging Bluetooth-related data, including paired devices and event logs.

- **Bug Fixes**
	- Enhanced error handling and logging in packet sniffing functionality.

- **Chores**
	- Updated version number in configuration file.
	- Modified dependencies in requirements by adding new packages.
	- Excluded virtual environment folder from project configuration.
	- Assigned code ownership for `bluetooth_details.py` and `bluetooth_logger.py`.
	- Updated documentation to reflect new Bluetooth data handling capabilities.
	- Removed outdated CSV editor configuration file.

### CODE REVIEW

File: CODE/bluetooth_details.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                f.write(f"Name: {device.Name}\n")
+                f.write(f"    Device ID: {device.DeviceID}\n")
+                f.write(f"    Description: {device.Description}\n")
+                f.write(f"    Manufacturer: {device.Manufacturer}\n")
+                f.write(f"    Status: {device.Status}\n")
+                f.write(f"    PNP Device ID: {device.PNPDeviceID}\n")
+                f.write("-" * 50)
+                f.write("\n\n")
+    except Exception as e:
+        log.error(f"Error: {e}")
```
[the remark this answers is not in the document]
Comment (ski-sketch) [reply CODE/bluetooth_details.py:48]: Not neccessary

File: CODE/bluetooth_details.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            for device in devices:
+                log.info(f"Name: {device.Name}")
+                f.write(f"Name: {device.Name}\n")
+                f.write(f"    Device ID: {device.DeviceID}\n")
+                f.write(f"    Description: {device.Description}\n")
+                f.write(f"    Manufacturer: {device.Manufacturer}\n")
+                f.write(f"    Status: {device.Status}\n")
+                f.write(f"    PNP Device ID: {device.PNPDeviceID}\n")
+                f.write("-" * 50)
+                f.write("\n\n")
```
[the remark this answers is not in the document]
Comment (ski-sketch) [reply CODE/bluetooth_details.py:46]: Thanks will do

File: CODE/bluetooth_details.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import win32com.client
+
+from logicytics import Log, DEBUG
+
+if __name__ == "__main__":
+    log = Log({"log_level": DEBUG})
```
[the remark this answers is not in the document]
Comment (ski-sketch) [reply CODE/bluetooth_details.py:6]: Nope, the project requires the logger to be that way

File: CODE/bluetooth_details.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    log.info("Fetching detailed info for Bluetooth devices...")
+    try:
+        locator = win32com.client.Dispatch("WbemScripting.SWbemLocator")
+        service = locator.ConnectServer(".", "root\\cimv2")
+
+        devices = service.ExecQuery("SELECT * FROM Win32_PnPEntity WHERE Name LIKE '%Bluetooth%'")
+
+        # Make sure it exists first
+        with open("Bluetooth Info.txt", "w", encoding="UTF-8") as f:
+            for device in devices:
```
[the remark this answers is not in the document]
Comment (ski-sketch) [reply CODE/bluetooth_details.py:37]: ? the keyword "w" in mode of open() auto-makes it

File: CODE/config.ini [hunk 5]
```diff
@@ -9,8 +9,8 @@ delete_old_logs = false
...
+version = 3.1.1
+files = "bluetooth_details.py, browser_miner.ps1, cmd_commands.py, dir_list.py, dump_memory.py, event_log.py, Logicytics.py, log_miner.py, media_backup.py, netadapter.ps1, packet_sniffer.py, property_scraper.ps1, registry.py, sensitive_data_miner.py, ssh_miner.py, sys_internal.py, tasklist.py, tree.ps1, vulnscan.py, wifi_stealer.py, window_feature_miner.ps1, wmic.py, _debug.py, _dev.py, _extra.py, logicytics\Checks.py, logicytics\Execute.py, logicytics\FileManagement.py, logicytics\Flag.py, logicytics\Get.py, logicytics\Logger.py, logicytics\__init__.py, VulnScan\tools\_study_network.py, VulnScan\tools\_test_gpu_acceleration.py, VulnScan\tools\_vectorizer.py, VulnScan\v2-deprecated\_generate_data.py, VulnScan\v2-deprecated\_train.py, VulnScan\v3\_generate_data.py, VulnScan\v3\_train.py"
```
[the remark this answers is not in the document]
Comment (ski-sketch) [reply CODE/config.ini:13]: Will do thanks

File: CODE/Logicytics.py [hunk 6]
```diff
@@ -313,10 +313,10 @@ def threaded_execution(execution_list_thread, index_thread):
...
     else:
         try:
-            EXECUTION_LIST = generate_execution_list()
-            for file in range(len(EXECUTION_LIST)):  # Loop through List
-                log.parse_execution(Execute.script(EXECUTION_LIST[file]))
-                log.info(f"{EXECUTION_LIST[file]} executed")
+            execution_list = generate_execution_list()
+            for file in range(len(execution_list)):  # Loop through List
+                log.parse_execution(Execute.script(execution_list[file]))
+                log.info(f"{execution_list[file]} executed")
```
[the remark this answers is not in the document]
Comment (ski-sketch) [reply CODE/Logicytics.py:319]: will do in next commit

File: CODE/bluetooth_details.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import win32com.client
+
+from logicytics import Log, DEBUG
+
+if __name__ == "__main__":
+    log = Log({"log_level": DEBUG})
+
+def get_bluetooth_device_details():
```
[the remark this answers is not in the document]
Comment (ski-sketch) [reply CODE/bluetooth_details.py:9]: Fixed

File: CODE/bluetooth_details.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            'Status': getattr(device, 'Status', 'Unknown'),
+            'PNP Device ID': getattr(device, 'PNPDeviceID', 'Unknown')
+        }
+        log.info(f"Retrieved device: {device_info['Name']}")
+        device_info_list.append(device_info)
+
+    return device_info_list
+
+
+def _write_device_info_to_file(devices, filename):
```
[the remark this answers is not in the document]
Comment (ski-sketch) [reply CODE/bluetooth_details.py:69]: No... lol

File: CODE/bluetooth_details.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    Returns:
+        None
+    """
+    log.info("Fetching detailed info for Bluetooth devices...")
+    try:
+        devices = _query_bluetooth_devices()
+        _write_device_info_to_file(devices, "Bluetooth Info.txt")
+    except Exception as e:
+        log.error(f"Error: {e}")
+        exit(1)
```
[the remark this answers is not in the document]
Comment (ski-sketch) [reply CODE/bluetooth_details.py:35]: Ehh, i think its fine

**Review by DefinetlyNotAI — APPROVED** [review]

**Comment by DefinetlyNotAI** [comment]
> **Actionable comments posted: 0**
> 
> 🧹 Nitpick comments (1)
> > README.md (1)> `216-217`: **Sweet addition! But let's make it even better! 🎮**
> > > Hey there! The new Bluetooth scripts look awesome, but we should probably mention that admin rights are needed to access the Windows registry and WMI stuff. That way, users won't get stuck wondering why it's not working!
> > > Here's how we could make it clearer:
> > > ```diff
> > > -| bluetooth_details.py     | Gets the PNP Device ID, Status, Manufacturer, Device ID, Name, Description of all paired bluetooth devices           |                          |
> > > -| bluetooth_logger.py      | Collect, log, and analyze Bluetooth-related data, by accessing the Windows registry and Event Viewer.                |                          |
> > > +| bluetooth_details.py     | Gets the PNP Device ID, Status, Manufacturer, Device ID, Name, Description of all paired bluetooth devices           | Requires admin rights    |
> > > +| bluetooth_logger.py      | Collect, log, and analyze Bluetooth-related data, by accessing the Windows registry and Event Viewer.                | Requires admin rights    |
> > > ```
> 
> 📜 Review details

The project won't even run without admin priv, so that is ok.

File: CODE/bluetooth_logger.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+# Function to execute PowerShell command and return results
+def run_powershell_command(command):
+    """Runs a PowerShell command and returns the output."""
+    try:
+        result = subprocess.run(["powershell", "-Command", command], capture_output=True, text=True)
+        return result.stdout.split('\n')
+    except Exception as e:
+        log.error(f"Error running PowerShell command: {e}")
+        return []
+
```
[the remark this answers is not in the document]
Comment (DefinetlyNotAI) [reply CODE/bluetooth_logger.py:65]: @ski-sketch After you remove some dependancies that I told you to remove ( `winreg` and `pywin32` ), push so I can fix these

**Review by DefinetlyNotAI — APPROVED** [review]

File: CODE/bluetooth_logger.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (DefinetlyNotAI) [reply CODE/bluetooth_logger.py:66]: @coderabbitai Any other suggestions to add?

[the remark this answers is not in the document]
Comment (DefinetlyNotAI) [reply CODE/bluetooth_logger.py]: Fixed

File: CODE/bluetooth_logger.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+                    file.write(f"{item}\n")
+            else:
+                file.write(f"{data}\n")
+            file.write(f"\n{'=' * 50}\n")
+    except Exception as e:
+        log.error(f"Error writing to file {filename}: {e}")
+
+
+# Function to collect paired Bluetooth devices
+def get_paired_bluetooth_devices():
```
[the remark this answers is not in the document]
Comment (DefinetlyNotAI) [reply CODE/bluetooth_logger.py]: Fixed

File: CODE/bluetooth_logger.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    except Exception as e:
+        log.error(f"Error parsing file transfer logs: {e}")
+
+    return transfer_logs
+
+
+# Main function to collect and log all Bluetooth data
+
+
+def main():
```
[the remark this answers is not in the document]
Comment (DefinetlyNotAI) [reply CODE/bluetooth_logger.py]: Fixed

File: .github/CODEOWNERS [hunk 13]
```diff
@@ -1,3 +1,5 @@
 * @DefinetlyNotAI
 wifi_stealer.py @ski-sketch
-packet_sniffer.py @ski-sketch
\ No newline at end of file
+packet_sniffer.py @ski-sketch
+bluetooth_details.py @ski-sketch
+bluetooth_logger.py @ski-sketch
```
[the remark this answers is not in the document]
Comment (DefinetlyNotAI) [reply .github/CODEOWNERS:5]: These files are part of Logicytics OSINT

**Comment by DefinetlyNotAI** [comment]
This pull request includes various updates and improvements across multiple files, with a focus on adding new Bluetooth-related functionality, enhancing existing methods, and updating configuration and documentation.

### Bluetooth Functionality Enhancements:
* Added `bluetooth_details.py` to retrieve and log detailed information about Bluetooth devices on the system.
* Added `bluetooth_logger.py` to collect, log, and analyze Bluetooth-related data by accessing the Windows registry and Event Viewer.

### Code Improvements:
* Refactored `threaded_execution` and other functions in `CODE/Logicytics.py` to use local variables (`execution_list`) instead of global variables (`EXECUTION_LIST`). [[1]](diffhunk://#diff-44a988332704ab86171610dcfe391d08df0f70f8177c3463e6e089aec9c5baa0L277-R282) [[2]](diffhunk://#diff-44a988332704ab86171610dcfe391d08df0f70f8177c3463e6e089aec9c5baa0L293-R300) [[3]](diffhunk://#diff-44a988332704ab86171610dcfe391d08df0f70f8177c3463e6e089aec9c5baa0L316-R319)
* Enhanced `get_flags` function in `CODE/Logicytics.py` to handle exceptions more accurately and improve readability.

### Configuration and Documentation Updates:
* Updated `.github/CODEOWNERS` to include new Bluetooth-related files.
* Updated `config.ini` to include new Bluetooth-related files and incremented version to 3.1.1.
* Updated `README.md` to document new Bluetooth-related functionalities.

### Miscellaneous Changes:
* Added `.venv` to excluded folders in `.idea/Logicytics.iml`.
* Removed obsolete `.idea/csv-editor.xml` configuration file.
* Fixed logical error in `_prompt_user` function in `CODE/_dev.py` to correctly handle "yes" and "y" responses.
* Improved file list handling in `dev_checks` function in `CODE/_dev.py`.
* Changed log level from `info` to `debug` in `print_packet_summary` function in `CODE/packet_sniffer.py`.
* Modified `visualize_graph` function in `CODE/packet_sniffer.py` to close the plot instead of showing it.
* Enhanced error logging in `main` function in `CODE/packet_sniffer.py` for invalid packet count or timeout values.

**Pull request merged**